### PR TITLE
ISPN-5345 Allow eviction for based on approximation of size compared to

### PR DESCRIPTION
**Pull request by wburns**
element count

* Added in size calculation for byte[] key/value pairs with LRU

https://issues.jboss.org/browse/ISPN-5345

### CODE REVIEW

File: commons/src/main/java/org/infinispan/commons/util/concurrent/jdk8backported/BoundedEquivalentConcurrentHashMapV8.java [hunk 1]
```diff
@@ -265,6 +265,96 @@
...
+         return dropBits + 8;
+      }
+      return size;
+   }
+
+   public static abstract class AbstractSizeCalculatorHelper<K, V> implements EntrySizeCalculator<K, V> {
+      // This is how large the object header info is
+      public final static int OBJECT_SIZE = sun.misc.Unsafe.ADDRESS_SIZE;
+      // This is how larege an object pointer is - note that each object
+      // has to reference it's class
```
Comment (tristantarrant) [inline commons/src/main/java/org/infinispan/commons/util/concurrent/jdk8backported/BoundedEquivalentConcurrentHashMapV8.java]: <grammar-nazi>its</grammar-nazi>

File: commons/src/main/java/org/infinispan/commons/util/concurrent/jdk8backported/BoundedEquivalentConcurrentHashMapV8.java [hunk 2]
```diff
@@ -265,6 +265,96 @@
...
+      }
+   }
+
+   public static final class NodeSizeCalculatorWrapper<K, V> extends AbstractSizeCalculatorHelper<K, V> {
+      private final EntrySizeCalculator<? super K, ? super V> calculator;
+      private final long nodeAverageSize;
+
+      public NodeSizeCalculatorWrapper(EntrySizeCalculator<? super K, ? super V> calculator) {
+         this.calculator = calculator;
+         // The node itself is an object and has a reference to it's class
```
Comment (tristantarrant) [inline commons/src/main/java/org/infinispan/commons/util/concurrent/jdk8backported/BoundedEquivalentConcurrentHashMapV8.java]: s/it's/its/

File: server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLReader_8_0.java [hunk 3]
```diff
@@ -798,6 +798,10 @@ private void parseEviction(XMLExtendedStreamReader reader, ModelNode cache, List
                     EvictionResource.MAX_ENTRIES.parseAndSetParameter(value, eviction, reader);
                     break;
                 }
+                case MEMORY_BASED_APPROXIMATION: {
+                    EvictionResource.MEMORY_BASED_APPROXIMATION.parseAndSetParameter(value, eviction, reader);
```
Comment (tristantarrant) [inline server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLReader_8_0.java]: Adding this is not enough, you also need to take care of it in CacheAdd

Comment (wburns) [inline server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLReader_8_0.java]: Ah yes thanks for the reminder :)

Comment (wburns) [inline server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLReader_8_0.java]: Added CacheAdd changes.

File: commons/src/main/java/org/infinispan/commons/util/concurrent/jdk8backported/BoundedEquivalentConcurrentHashMapV8.java [hunk 4]
```diff
@@ -265,6 +265,96 @@
...
+      long calculateSize(K key, V value);
+   }
+
+   private static final Long IGNORE_BOTTOM_3_BITS = Long.MAX_VALUE << 3;
+
+   private static long roundUpToNearest8(long size) {
+      // If there is a remainder we have to round up
+      if ((size & 0x7) != 0) {
+         long dropBits = size & IGNORE_BOTTOM_3_BITS;
+         return dropBits + 8;
```
Comment (danberindei) [inline commons/src/main/java/org/infinispan/commons/util/concurrent/jdk8backported/BoundedEquivalentConcurrentHashMapV8.java]: I think `(size + 7) & ~0x7` would look better ;)

Comment (wburns) [inline commons/src/main/java/org/infinispan/commons/util/concurrent/jdk8backported/BoundedEquivalentConcurrentHashMapV8.java]: Ah good point :+1:

**Comment by danberindei** [comment]
Looks good, but I think you should replace the `memory-based : boolean` attribute with a `max-entries-memory : long` attribute that would override `max-entries`.

Also, did you see any change in performance with the memory-based eviction enabled?

**Comment by wburns** [comment]
In the quick tests I ran it was negligible - I can try running some more tests on it though next week possibly.

Yeah I wasn't quite sure what I wanted to do with the parser.  @tristantarrant do you have any preference (you seem to to be the parser master)?

**Comment by tristantarrant** [comment]
Yes, I do not like overloading "maxEntries" to also mean memory size. Also, since you cannot have mutually exclusive attributes in XSD, probably it's best to add an additional sub-element to the <eviction> element and deprecate the maxEntries attribute in XML. I'd also like to be able to specify readable units there. Are max entries and max memory really mutually exclusive ?

<eviction>
<max-entries>100</max-entries>
</eviction>

<eviction>
<max-memory>100m</max-memory>
</eviction>

**Comment by wburns** [comment]
In the current implementation they are mutually exclusive.  Do we think a user will want to configure both?  I can try to see what it would take to add that (my guess is it may not be that much additional work)

**Comment by tristantarrant** [comment]
Ok, then my statement about mutually exclusive schema stands

**Comment by tristantarrant** [comment]
I also updated my comment above so that the XML doesn't get squished

**Comment by wburns** [comment]
@tristantarrant Do you think we can integrate without the parser changes and I can work on them next week?  I am not quite sure how everything should look yet.

**Comment by wburns** [comment]
Updated the 8 byte rounding.  I can try to get the parser done when I get back.

**Comment by danberindei** [comment]
@wburns Unless you think rebasing will be a pain, I'd keep everything in one PR.

**Comment by wburns** [comment]
@danberindei I was more saying because we have a build this Friday and I won't be able to get all the parser changes in by then.

**Comment by tristantarrant** [comment]
I'll take over this PR, and add the appropriate parsing

**Comment by tristantarrant** [comment]
Replacing this with https://github.com/infinispan/infinispan/pull/3492

**Pull request closed without merging**